agent/consul/acl.go

```diff
@@ -1592,6 +1592,13 @@ func vetDeregisterWithACL(rule acl.Authorizer, subj *structs.DeregisterRequest,
 	// We don't apply sentinel in this path, since at this time sentinel
 	// only applies to create and update operations.
 
+	// Allow service deregistration if the token has write permission on the agent.
+	// This accounts for cases where the agent no longer has a token with write permission
+	// on the service to deregister it.
+	if rule.NodeWrite(subj.Node, nil) {
```

Comment:
Random thought: are we sure `subj.Node` is always populated? Are there cases where we send a `DeregisterRequest` for a whole service not attached to a specific node? In this case would this behave correctly?

Reply:
yes, `vetDeregisterWithACL` is only called in `Catalog.Deregister` and theres a check making sure Node is not empty